### PR TITLE
fix firefox issue in windows

### PR DESCRIPTION
**Pull request by 7aman**
This PR adds a `persepolis/msghost.py` as a middle man to transfer messages from browsers to persepolis.  It just affects windows users and has nothing to do with other operating systems.  
It also skips sending a response after getting browser messages (for Windows only). It is not a vital part and I will fix it later.  
After this PR merged, I will update [persepolis-windows-package-build](https://github.com/persepolisdm/persepolis-windows-package-build) and make a PR to adopt this changes.

### CODE REVIEW

File: persepolis/scripts/persepolis.py [hunk 1]
```diff
@@ -186,19 +186,21 @@ def setPersepolisColorScheme(self, color_scheme):
...
         msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
         msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
-
-    # Send message to browsers plugin
-    message = '{"enable": true, "version": "1.85"}'.encode('utf-8')
-    sys.stdout.buffer.write((struct.pack('i', len(message))))
-    sys.stdout.buffer.write(message)
-    sys.stdout.flush()
+    else:
+        # TODO: fix windows error
```
Comment (JafarAkhondali) [inline persepolis/scripts/persepolis.py:197]: This line is in `else` part of  `if os_type == OS.WINDOWS:` so it shouldn't throw error for Windows !?

Comment (7aman) [reply persepolis/scripts/persepolis.py:197]: Yes.   
It could be in another `if os_type != OS.WINDOWS:` but this version also does the work.

Comment (JafarAkhondali) [reply persepolis/scripts/persepolis.py:197]: Oh ok i misunderstood the comment.

File: persepolis/msghost.py [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+## - message from Chrome will be dropped.
+## As a temporary solution, message is parsed here and saved to a binary file.
+## Then file is passed to pdm.
+home_address = str(os.path.expanduser("~"))
+nhm_request = os.path.join(
+    home_address, r'AppData\Local\persepolis_download_manager\nhm_request.bin')
+raw_size = sys.stdin.buffer.read(4)  
+length = struct.unpack('@I', raw_size)[0]
+raw_data = sys.stdin.buffer.read(length)
+with open(nhm_request, "w+b") as msg:
```
Comment (JafarAkhondali) [inline persepolis/msghost.py]: Why write in a file? Can't we use a variable instead and prevent IO?

File: persepolis/scripts/persepolis.py [hunk 3]
```diff
@@ -186,19 +186,21 @@ def setPersepolisColorScheme(self, color_scheme):
...
+   
     # Platform specific configuration
     if os_type == OS.WINDOWS:
         # Set the default I/O mode to O_BINARY in windows
         import msvcrt
         msvcrt.setmode(sys.stdin.fileno(), os.O_BINARY)
         msvcrt.setmode(sys.stdout.fileno(), os.O_BINARY)
-
-    # Send message to browsers plugin
-    message = '{"enable": true, "version": "1.85"}'.encode('utf-8')
```
Comment (JafarAkhondali) [inline persepolis/scripts/persepolis.py:198]: So if platform is not windows, there is no response from Persepolis?

File: persepolis/msghost.py (hunk 2 above)
Comment (7aman) [reply persepolis/msghost.py]: The idea is to redirect or simulate `sys.stdin` that is in binary format.  
I do not know how to do that with a variable. :smile:

File: persepolis/scripts/persepolis.py (hunk 3 above)
Comment (7aman) [reply persepolis/scripts/persepolis.py:198]: If os is windows, skip sending the response.

File: persepolis/scripts/persepolis.py [hunk 4]
```diff
@@ -186,19 +186,21 @@ def setPersepolisColorScheme(self, color_scheme):
...
-    sys.stdout.buffer.write((struct.pack('i', len(message))))
-    sys.stdout.buffer.write(message)
-    sys.stdout.flush()
+    else:
+        # TODO: fix windows error
+        # This section causes an "OSError: [Errno 22] Invalid argument" in windows, skip for now.
+        # Send message to browsers plugin
+        message = '{"enable": true, "version": "1.85"}'.encode('utf-8')
+        sys.stdout.buffer.write((struct.pack('i', len(message))))
```
Comment (7aman) [inline persepolis/scripts/persepolis.py:201]: this line throws an `OSError` in Windows. I could not figure out how to resolve this yet.

Comment (JafarAkhondali) [reply persepolis/scripts/persepolis.py:201]: Ok, but this is important. We'll need to fix it

File: persepolis/msghost.py (hunk 2 above)
Comment (7aman) [reply persepolis/msghost.py]: i fixed it by using communicate method.

**Comment by alireza-amirsamimi** [comment]
@7aman 
Hello
Sorry for late answer. Thank you very much.
Is this ready to use or test?

**Comment by 7aman** [comment]
Hi.
I am not sure anymore. I lost my interest and I don't remember all the details about what was I doing.

From reviews above, I remember this PR has still some issue for sending messages to add-on but I doubt that it is a very serious problem. I was just interested to receive add-on messages and launch Persepolis if it was not running. And I think I fixed it. 


Also it is related to this issue. It affects the way Persepolis is freezed for windows.
https://github.com/persepolisdm/persepolis-windows-package-build/pull/3
I suggest that Persepolis team must provide an easy and clear instruction to run Persepolis from source code on windows platform. It is necessary to have an editable installation to debug and make changes.

**Pull request closed without merging**